### ISSUE TITLE
Keep validation result in state

### ISSUE THREAD

**Issue by ivan-kleshnin** [jurassix/react-validation-mixin#2]
Currently, validation may be called multiple times with the same data. For example, `isValid` may be called on submit and `getValidationMessages(<field-name>)` may be called n times for each field to get and render error messages.

```javascript
isValid: function(field) {
  return ValidationFactory.isValid(this.validate(), field);
}

getValidationMessages: function(field) {
  return ValidationFactory.getValidationMessages(this.validate(), field);
},
```

Instead of returning validation messages it's probably better to keep them in state.
This link may help: [Reference implementation](http://stackoverflow.com/questions/21854938/using-mixins-vs-components-for-code-reuse-in-facebook-react)

**Comment by jurassix** [comment]
Thanks for the report.

Yup, I've been playing with different options to correct the current performance overhead of validating single fields verse validating all fields. I like the idea of caching the result set; I don't want to override any of the component lifecycle methods unless absolutely necessary.

**Comment by ivan-kleshnin** [comment]
> I don't want to override any of the component lifecycle methods unless absolutely necessary.

I'ts common scenario to provide self-validation (in terms of required properties, not *that* validtion) in `componentDidMount`, for example. I saw this use-case a lot from trusted React sources.
React mixin's methods don't technically override each other, they chain. That's cool.
So I think mixins which override "component lifecycle methods" are absolutely ok.

**Comment by jurassix** [comment]
I didn't know that mixin's chained; good information.

**Comment by jurassix** [comment]
This is resolved and published to npm in version 2.1.0

I'm maintaining a cache of validation results and invalidating that cache automatically during the ComponentWillUpdate lifecycle. I've also exposed a manual clearValidations() api if you need to force a revalidation during a render cycle. 

Validations are also lazily initialized; we only validate on the first call to isValid or getValidationsMessages. This now ensures only the first call has a processing overhead and all subsequent render calls are pulling from cache

**Comment by ivan-kleshnin** [comment]
Ah, ok. I always look into **Releases** tab and it somehow misses last ones. 
Final tag it shows is **v2.0.2** :( 
I think you should give it a look.

**Comment by jurassix** [comment]
Fixed

**Comment by ivan-kleshnin** [comment]
I think I have a strong argument for `this.state` usage.

Besides of validation error messages there are other errors. Like "email is already used".
Techically, it's not a validation error. And Joi or other validation libraries can't help with that, as such errors are app specific. But, as developer, you'll probably want to show them right there along with validation errors. So we need a way to merge custom error messages (obtained from custom "validation" or other similar mechanics) with a **react-validation-mixin** validation messages.

If library uses `this.state.errors` it's simple. Just push whatever you want. If it's cached private var – it's both hard and hacky. It's quite possible to add secondary error messages. But it violates `single source of truth` principle for UI elements.

**Comment by ivan-kleshnin** [comment]
Also, did you saw this? React documentation is awesome:
https://github.com/facebook/react/blob/0c6bee049efb63585fb88c995de788cefc18b789/src/core/ReactCompositeComponent.js#L47

**Comment by jurassix** [comment]
Thanks for the argument and research. I agree with you on using state to store the errors, caching on 'this' didn't seem right for React. I'll correct this in the next release.

**Comment by jurassix** [comment]
fixed in version 3.0.0

validation errors are now stored on the components state: `this.state.errors`

**Comment by ivan-kleshnin** [comment]
Wow, You're blazingly fast  :)
